src/ringbuffer_trait.rs

```diff
@@ -1,53 +1,217 @@
-    fn iter(&'a self) -> Self::Iter;
+    #[cfg(not(tarpaulin_include))]
+    #[inline]
+    fn iter(&self) -> RingBufferIterator<T, Self> {
+        RingBufferIterator {
+            obj: &self,
+            index: 0,
+            phantom: Default::default(),
+        }
+    }
 
     ///  Creates a mutable iterator over the buffer starting from the latest push.
-    fn iter_mut(&'b mut self) -> Self::IterMut;
-
-    /// Converts the buffer to an vector.
-    #[cfg(feature = "alloc")]
-    fn as_vec(&self) -> Vec<&T>;
+    #[cfg(not(tarpaulin_include))]
+    #[inline]
+    fn iter_mut(&mut self) -> RingBufferMutIterator<T, Self> {
+        RingBufferMutIterator {
+            obj: self,
+            index: 0,
+            phantom: Default::default(),
+        }
+    }
 
     /// Converts the buffer to an vector.
     #[cfg(feature = "alloc")]
     fn to_vec(&self) -> Vec<T>
     where
         T: Clone,
     {
-        self.as_vec().into_iter().cloned().collect()
+        self.iter().cloned().collect()
     }
+
+    /// Returns true if elem is in the ringbuffer
+    fn contains(&self, elem: &T) -> bool
+    where
+        T: PartialEq,
+    {
+        self.iter().any(|i| i == elem)
+    }
+}
+
+pub struct RingBufferIterator<'rb, T: 'static + Default, RB: RingBuffer<T>> {
+    obj: &'rb RB,
+    index: usize,
+    phantom: PhantomData<T>,
+}
+
+impl<'rb, T: 'static + Default, RB: RingBuffer<T>> Iterator for RingBufferIterator<'rb, T, RB> {
+    type Item = &'rb T;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.obj.len() {
+            let res = self.obj.get(self.index as isize);
+            self.index += 1;
+            res
+        } else {
+            None
+        }
+    }
+}
+
+pub struct RingBufferMutIterator<'rb, T: 'static + Default, RB: RingBuffer<T>> {
+    obj: &'rb mut RB,
+    index: usize,
+    phantom: PhantomData<T>,
+}
+
+impl<'rb, T: 'static + Default, RB: RingBuffer<T>> Iterator for RingBufferMutIterator<'rb, T, RB> {
+    type Item = &'rb mut T;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.obj.len() {
+            let res: Option<&'_ mut T> = self.obj.get_mut(self.index as isize);
+            self.index += 1;
+
+            // Safety:
+            // This mem transmute is extending the lifetime of the returned value.
+            // This is necessary because the rust borrow checker is too restrictive in giving out mutable references.
+            // It thinks the iterator can give out a mutable reference, while it's also possible to mutably borrow
+            // `obj` in the RingBufferMutIterator struct. This is however *never* possible because it's a private field
+            // Unfortunately this is a limitation of the rust compiler. It's well explained here:
+            // http://smallcultfollowing.com/babysteps/blog/2013/10/24/iterators-yielding-mutable-references/
```

Comment:
Maybe an idea to restrict access to the internal `obj`  is to place the RingBufferMutIterator (maybe together with the RingBufferIterator) into a seperate module, so that access to the obj is made even less likely / more difficult